lib/TPP/CombineXsmmPass.cpp

```diff
@@ -0,0 +1,194 @@
+
+  int64_t m = memrefC.cast<ShapedType>().getShape()[0];
+  int64_t n = memrefC.cast<ShapedType>().getShape()[1];
+  int64_t k = memrefA.cast<ShapedType>().getShape()[2];
+
+  auto ldaDim = xsmm::utils::getLeadingDim(memrefA, /*pos=*/1);
+  if (failed(ldaDim)) {
+    return failure();
+  }
+  int64_t lda = *ldaDim;
+
+  auto ldbDim = xsmm::utils::getLeadingDim(memrefB, /*pos=*/1);
+  if (failed(ldbDim)) {
+    return failure();
+  }
+  int64_t ldb = (vnni::utils::isInVnniLayout(memrefB.cast<MemRefType>()))
+                    ? *ldbDim / *vnni::utils::getVnniBlockingFactor(memrefB)
+                    : *ldbDim;
+
+  auto ldcDim = xsmm::utils::getLeadingDim(memrefC);
+  if (failed(ldcDim)) {
+    return failure();
+  }
+  int64_t ldc = *ldcDim;
+
+  // If we are dealing with a BRGEMM we need to pass two extra dimensions:
+  // - strideA and strideB that represent the stride between different GEMM
+  // in BRGEMM.
+  int64_t strideA = lda * m;
+  int64_t strideB = ldb * k;
+  return DenseI64ArrayAttr::get(
+      rewriter.getContext(),
+      ArrayRef<int64_t>{m, n, k, lda, ldb, ldc, strideA, strideB});
+}
+
+ArrayAttr getUnaryFlags(RewriterBase &rewriter) {
+  return rewriter.getArrayAttr(
+      xsmm::UnaryFlagsAttr::get(rewriter.getContext(), xsmm::UnaryFlags::NONE));
+}
+
+ArrayAttr getBinaryFlags(RewriterBase &rewriter, xsmm::BinaryOp addOp) {
+  auto binaryInputType = addOp.getOperand(2).getType().cast<MemRefType>();
+  auto outputType = addOp.getOperand(3).getType().cast<MemRefType>();
+  return rewriter.getArrayAttr(xsmm::BinaryFlagsAttr::get(
+      rewriter.getContext(),
+      xsmm::utils::getBinaryBCast(binaryInputType, outputType,
+                                  /*operandNumber=*/0)));
+}
+
+xsmm::BinaryKindAttr getBinaryKind(RewriterBase &rewriter) {
+  return xsmm::BinaryKindAttr::get(rewriter.getContext(),
+                                   xsmm::BinaryKind::ADD);
+}
+
+xsmm::UnaryKindAttr getUnaryKind(RewriterBase &rewriter) {
+  return xsmm::UnaryKindAttr::get(rewriter.getContext(), xsmm::UnaryKind::RELU);
+}
+
+static ArrayAttr getGemmFlags(RewriterBase &rewriter, xsmm::BrgemmOp opTy) {
+  auto memrefB = opTy.getOperand(2).getType().cast<MemRefType>();
+  xsmm::GemmFlagsAttr gemmFlag =
+      (vnni::utils::isInVnniLayout(memrefB))
+          ? xsmm::GemmFlagsAttr::get(rewriter.getContext(),
+                                     xsmm::GemmFlags::VNNI_B)
+          : xsmm::GemmFlagsAttr::get(rewriter.getContext(),
+                                     xsmm::GemmFlags::NONE);
+  return rewriter.getArrayAttr(gemmFlag);
+}
+
+struct CombineXsmmOp : public OpRewritePattern<xsmm::BrgemmOp> {
+  using OpRewritePattern<xsmm::BrgemmOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(xsmm::BrgemmOp brgemmOp,
+                                PatternRewriter &rewriter) const override {
+    auto brgemmResultOperand = brgemmOp.getOperand(3);
+
+    for (auto use = brgemmResultOperand.user_begin();
```

Comment:
Is this actually safe? We need to be more strict with the users, what if you have:
```
xsmm.add(B, A)
xsmm.sub(C, A)
xsmm.brgemm(A ..)
```
this should not fuse as the sub writes into A. Can you please add some tests along this line?

Reply:
will address in a a subsequent commit

Reply:
This commit should be safe enough to not handle untested cases.

Reply:
The pattern here is in memrefs, so you need the producer of the result operand's user list. This can be an `alloc`, `alloca`, `region_op`, `subview`, etc.

As @chelini said, we need to make sure the `brgemm` is the first operation on the operand (other than `zero`) and the other "known" ops follow. We need to make sure the graph is consistent, too.

```mlir
  %a = memref.subview ...
  %b = memref.subview ...
  %c = memref.subview ...

  // optional
  xsmm.unary zero (%c)

  // brgemm, %c needs to be outs
  xsmm.brgemm (%a, %b) -> (%c)

  // binary in-place (%bias can be a broadcast, cannot be %a or %b)
  xsmm.binary add (%c, %bias) -> (%c)

  // binary out-of-place (%bias can be a broadcast, cannot be %a or %b)
  %add = memref.alloc
  xsmm.binary add (%c, %bias) -> (%add)

  // unary in-place
  xsmm.unary relu (%c) -> (%c)

  // unary out-of-place
  %relu = memref.alloc
  xsmm.unary relu (%c) -> (%relu)

  // User or %c / %add / %relu
  return ...
```

Reply:
For the record, initially we should not worry about `allocs`. We can clean these up on an earlier pass, or worry about them after we create the first implementation. Right now, let's assume all is in-place.